### PR TITLE
gh-124697: Represent inlined comprehensions as subscopes in the symbol table

### LINKED ISSUE #124697: refactor the implementation of inlined comprehensions
The current implementation of comprehension inlining is a pile of special cases that is quite difficult to understand and debug. The root cause of this is that it attempts to flatten the inlined scope into the enclosing scope in the symbol table, to avoid the compiler needing to generally understand the concept of inlined sub-scopes, but this flattening loses information and requires a number of special cases. And then the compiler implementation of inlining modifies the flattened scope as it traverses the inlined comprehension(s) in order to ensure correct handling of the names within the inlined sub-scope; ideally the compiler would not modify the symbol table as it generates code.

A better implementation would leave the scopes separate in the symbol table, losing no information but just marking the comprehension scope as an inlined sub-scope, and then add explicit support for inlined sub-scopes in the compiler (with appropriate query methods in the symtable.)

### Linked PRs
* gh-156819

### PR DESCRIPTION
**Pull request by iritkatriel**
Resolves #124697

Inlined comprehensions are now represented in the symbol table as block of a new type `InlinedComprehensionBlock`, which is a subscope of the enclosing scope (not a separate compilation unit). 

This moves the complexity of compiling inlined comprehensions from codegen to the symbol table
construction. 

It removes the smell of the compiler modifying the symbol table in codegen.

* Issue: gh-124697

### CODE REVIEW

File: Python/symtable.c [hunk 1]
```diff
@@ -840,71 +917,103 @@ inline_comprehension(PySTEntryObject *ste, PySTEntryObject *comp,
...
+    for (Py_ssize_t i = 0; i < PyList_GET_SIZE(comp->ste_children); i++) {
+        PySTEntryObject *child = (PySTEntryObject *)PyList_GET_ITEM(comp->ste_children, i);
+        if (child->ste_type != InlinedComprehensionBlock) {
+            continue;
+        }
+        PyObject *child_free = get_freevar_names(child);
+        if (child_free == NULL) {
+            return 0;
+        }
+        int ok = finalize_inlined_comprehension(ste, child, child_free,
```
Comment (carljm) [inline Python/symtable.c]: I think nested inlined comprehensions need to be finalized against their immediate parent here, rather than continuing to use the original class scope.

With:
```python
class C:
    x = 99
    result = [[x for _ in (0,)] for x in (42,)]
```

this PR segfaults while compiling; main returns `[[42]]`. The inner `x` stays `FREE` because `class_binds_free_name()` sees `C.x`, instead of falling through to the outer comprehension's `CELL`.

File: Python/symtable.c [hunk 2]
```diff
@@ -840,71 +917,103 @@ inline_comprehension(PySTEntryObject *ste, PySTEntryObject *comp,
...
+    for (Py_ssize_t i = 0; i < PyList_GET_SIZE(to_remove); i++) {
+        PyObject *name = PyList_GET_ITEM(to_remove, i);
+        if (PyDict_DelItem(comp->ste_symbols, name) < 0) {
+            goto error;
+        }
     }
-    if (remove_dunder_cond_annotations && PyDict_DelItemString(comp->ste_symbols, "__conditional_annotations__") < 0) {
-        return 0;
+    Py_CLEAR(to_remove);
+    for (Py_ssize_t i = 0; i < PyList_GET_SIZE(comp->ste_children); i++) {
```
Comment (carljm) [inline Python/symtable.c:920]: Could we avoid recursively walking the already-hoisted descendants here? `analyze_block()` splices each inlined comprehension's children into ancestor lists, so a depth-n chain is revisited roughly `2**n` times.

On a debug build, the following took about 0.74s on this PR versus 0.0014s on the base; depth 26 took several seconds:
```python
source = "[" * 24 + "0" + " for x in ()]" * 24
symtable.symtable(source, "?", "exec")
```

A recursive public `get_children()` walk also visits seven comprehension entries for only three lexical comprehensions. Maybe finalization needs a lexical-only child list or a visited marker.

File: Python/codegen.c [hunk 3]
```diff
@@ -5069,7 +5065,7 @@ codegen_comprehension(compiler *c, expr_ty e, int type,
     IterStackPosition iter_state;
     if (is_inlined) {
         VISIT(c, expr, outermost->iter);
-        if (push_inlined_comprehension_state(c, loc, entry, &inline_state)) {
+        if (push_inlined_comprehension_state(c, loc, entry, &inline_state) < 0) {
```
Comment (carljm) [inline Python/codegen.c:5073]: What restores `inline_state.saved_ste` if setup fails after `_PyCompile_EnterInlinedComprehensionScope()` succeeds? `codegen_push_inlined_comprehension_locals()` can fail while allocating `pushed_locals` or emitting instructions, and this jumps straight to `error`, which only clears `pushed_locals`.

File: Python/symtable.c [hunk 4]
```diff
@@ -840,71 +917,103 @@ inline_comprehension(PySTEntryObject *ste, PySTEntryObject *comp,
...
+            continue;
+        }
+        if (scope == GLOBAL_IMPLICIT || scope == GLOBAL_EXPLICIT) {
+            continue;
+        }
+        int keep = class_binds_free_name(ste, k, comp_flags);
+        if (keep < 0) {
+            goto error;
+        }
+        if (!keep) {
```
Comment (carljm) [inline Python/symtable.c]: I'm not sure we want the sparse delta to erase `USE` here.

```python
def outer(x):
    def inner():
        return [x for y in ()]
```

On main, `inner.lookup("x")` is `FREE|USE`. Here it is `FREE` with `is_referenced() == False`, while the comprehension's `lookup("x")` raises `KeyError`; no public table reports that `x` is referenced.

File: Doc/library/symtable.rst [hunk 5]
```diff
@@ -57,6 +57,14 @@ Examining Symbol Tables
 
       Used for the symbol table of a class.
 
+   .. attribute:: INLINED_COMPREHENSION
+      :value: "inlined comprehension"
+
+      Used for the symbol table of a list, set or dict comprehension that
```
Comment (carljm) [inline Doc/library/symtable.rst]: Could we document the sparse lookup and child-hoisting behavior here, not only mention the new type? Inlined entries omit names whose scope matches the enclosing table, so `lookup()` can raise `KeyError` for a name used in the comprehension; nested entries can also appear both under the comprehension and hoisted into an ancestor's `get_children()` list. Those differences matter to existing `symtable` consumers.

File: Python/symtable.c [hunk 6]
```diff
@@ -801,37 +813,102 @@ is_free_in_any_child(PySTEntryObject *entry, PyObject *key)
...
 }
 
+/* True if name is FREE in the comprehension and bound in the enclosing class.
+ * Those names are kept in the compressed delta so lookup does not treat them
+ * as class locals. */
 static int
-inline_comprehension(PySTEntryObject *ste, PySTEntryObject *comp,
-                     PyObject *scopes, PyObject *comp_free,
-                     PyObject *inlined_cells)
+class_binds_free_name(PySTEntryObject *ste, PyObject *name, long comp_flags)
```
Comment (carljm) [inline Python/symtable.c]: This comment / function name are low on context. The function name doesn't suggest that this has anything to do with comprehensions, but the comment assumes the reader knows that it does, and knows what a "compressed delta" is.

In general it would be really nice to have somewhere in a comment a more comprehensive explanation of how comprehension inlining is intended to work. (Yes, that would have been good in the previous version too!)

**Review by carljm — COMMENTED** [review]
This looks great! Thank you for working on this ❤️

File: Python/symtable.c (hunk 2 above)
Comment (iritkatriel) [reply Python/symtable.c:920]: Oh! good catch. I think the splice is no longer necessary.

File: Python/symtable.c (hunk 6 above)
Comment (iritkatriel) [reply Python/symtable.c]: I agree. The name is ambiguous/vague because I changed it in order to use this function in two places where the same check is applied in other contexts. That makes it worse because the comment is stale. Let me try to sort this out.

Comment (iritkatriel) [reply Python/symtable.c]: This whole thing about removing names from the comprehension symtable should probably go now. We want that symtable to reflect reality, and the reality is that these symbols are free in the comprehension. It's the lookup that needs to be fixed.

**Review by JelleZijlstra — COMMENTED** [review]
Very nice! This also fixes https://github.com/python/cpython/issues/121377 and https://github.com/python/cpython/issues/156664, but doesn't add tests for their reproducers. Can you add those tests?

I have a PR for the latter almost ready at #156691, but I'm happy to retire it in favor of this PR. @carljm @iritkatriel what do you think about backporting either my change or others? This PR feels a bit much to backport, but we may want to backport fixes for some of the bugs. 156691 is also a somewhat invasive change though.

Codex found that this now fails:

```python
import sys

def outer(x):
    def inner():
        return [(lambda: x, dict(**sys._getframe().f_locals))
                for x in x]
    return inner()

outer([1])
```

With `TypeError: dict() got multiple values for keyword argument 'x'.`. #156691 has a larger change to framelocalsproxy to deal with this sort of thing; you may want to incorporate its approach.

**Comment by JelleZijlstra** [comment]
Another relevant bug is #156091 but that one still crashes under this PR.

**Comment by carljm** [comment]
I'd be cautious about backporting an invasive fix for https://github.com/python/cpython/issues/156664. It's clearly a rarely-encountered edge case, given that nobody other than PyPy's test suite discovered the bug since 3.12, and the risk of introducing new bugs in an invasive fix seems high (particularly without this refactor in place.) I think "better the bugs you know than the ones you don't" applies in this case.

**Comment by iritkatriel** [comment]
Thanks. 

I pushed a fix for the frame locals bug. 

I have a fix for https://github.com/python/cpython/pull/156691  too, but it can go in a different PR because it's an orthogonal problem.

File: Python/compile.c [hunk 7]
```diff
@@ -1057,120 +1160,23 @@ _PyCompile_ResolveNameop(compiler *c, PyObject *mangled, int scope,
...
-                        return ERROR;
-                    }
-                }
-                Py_XDECREF(orig);
-            }
-        }
-    }
+    assert(state->saved_ste == NULL);
+    state->saved_ste = c->u->u_ste;
+    c->u->u_ste = (PySTEntryObject *)Py_NewRef(entry);
```
Comment (carljm) [inline Python/compile.c]: Could we resolve the implicit receiver for optimized `super()` calls through the enclosing function's symbol table?

This now aborts compilation in a debug build:

```python
class A:
    def f(self):
        return 42

class B(A):
    def f(self):
        return [super().f() for _ in (0,)]

print(B().f())
```

The exact base returns `[42]`. Here, `load_args_for_super()` synthesizes a load of `self`, which isn't present in the comprehension's symbol table, so `codegen_nameop()` hits its scope assertion. Renaming the receiver to `_self` gets past the assertion but raises `SystemError: no locals found` at runtime.

Can we cover the implicit receiver case, including nested comprehensions, when adapting this lookup?

File: Objects/frameobject.c [hunk 8]
```diff
@@ -380,16 +380,32 @@ framelocalsproxy_keys(PyObject *self, PyObject *Py_UNUSED(ignored))
...
+    }
 
     for (int i = 0; i < co->co_nlocalsplus; i++) {
         if (framelocalsproxy_hasval(frame->f_frame, co, i)) {
             PyObject *name = PyTuple_GET_ITEM(co->co_localsplusnames, i);
+            int found = PySet_Contains(seen, name);
+            if (found < 0) {
+                goto error;
+            }
+            if (found) {
```
Comment (carljm) [inline Objects/frameobject.c:397]: I think if we do this, we should apply the same deduplication to `items()`, `values()`, and `__len__()`. With the separate comprehension-cell and enclosing-free slots, these now disagree with `keys()`:

```python
import sys

def outer(x):
    def inner():
        return [(lambda: x,
                 len(sys._getframe().f_locals),
                 sys._getframe().f_locals.keys(),
                 sys._getframe().f_locals.items())
                for x in x]
    return inner()[0][1:]

print(outer([1, 2]))
```

Main returns `(1, ['x'], [('x', 1)])`; this revision returns `(2, ['x'], [('x', 1), ('x', [1, 2])])`. `values()` likewise includes both values, and `dict(f_locals.items())` silently selects the enclosing list instead of the iteration value returned by `dict(f_locals)`.

Let's extend the new frame-locals test to check that all of these operations agree.

File: Python/symtable.c [hunk 9]
```diff
@@ -801,110 +816,126 @@ is_free_in_any_child(PySTEntryObject *entry, PyObject *key)
...
-        // drop_class_free) unless children scopes need it
+        // These names are not allowed to be free through a class (see
+        // drop_class_free) unless a nested child needs them. Keep FREE
+        // on this table; compile treats the load as a global.
         if (scope == FREE && ste->ste_type == ClassBlock &&
-                (_PyUnicode_EqualToASCIIString(k, "__class__") ||
-                 _PyUnicode_EqualToASCIIString(k, "__classdict__") ||
-                 _PyUnicode_EqualToASCIIString(k, "__conditional_annotations__"))) {
-            scope = GLOBAL_IMPLICIT;
+                _PyST_IsClassClosureName(k)) {
```
Comment (carljm) [inline Python/symtable.c:875]: I think this check now counts an inlined child as needing a real class closure, even though its special-name references compile as global lookups.

For example:

```python
class Meta(type):
    def __new__(mcls, name, bases, ns):
        ns.pop('__classcell__', None)
        return type.__new__(mcls, name, bases, ns)

class C(metaclass=Meta):
    result = [[super for _ in (0,)] for _ in (0,)]
```

Main creates `C` without supplying a `__classcell__`. This PR adds the cell and raises `RuntimeError: __class__ not set defining 'C'`. A single comprehension still succeeds.

Can we traverse through inlined children to determine whether an actual nested compilation unit needs the cell?

File: Python/symtable.c [hunk 10]
```diff
@@ -1262,24 +1285,20 @@ analyze_block(PySTEntryObject *ste, PyObject *bound, PyObject *free,
...
+        // enclosing scope. Nested ones are finalized recursively against
+        // their immediate parent.
         if (!analyze_child_block(entry, newbound, newfree, newglobal,
                                  type_params, new_class_entry, &child_free))
         {
             goto error;
         }
-        if (inline_comp) {
-            if (!inline_comprehension(ste, entry, scopes, child_free, inlined_cells)) {
+        if (entry->ste_type == InlinedComprehensionBlock && ste->ste_type != InlinedComprehensionBlock) {
```
Comment (carljm) [inline Python/symtable.c:1296]: Can we preserve the fast-local optimization when a variable is only referenced by other inlined comprehensions?

```python
def f(n):
    return [[x for _ in range(20)] for x in range(20)]
```

Deferring finalization until the enclosing scope is no longer inlined lets `analyze_cells()` promote `x` before those references are handled. Main has no cells or dereference instructions here; this revision has `co_cellvars == ('x',)`, two `MAKE_CELL` instructions, and `STORE_DEREF`/`LOAD_DEREF`.

Small `timeit` measurements on matching debug builds showed this case around 11–17% slower. I haven't measured release-build impact.

Maybe this is fine? But it seems like it should be possible to keep this refactor behavior-preserving.

File: Doc/library/symtable.rst [hunk 11]
```diff
@@ -57,6 +57,14 @@ Examining Symbol Tables
 
       Used for the symbol table of a class.
 
+   .. attribute:: INLINED_COMPREHENSION
```
Comment (carljm) [inline Doc/library/symtable.rst:60]: Nit: could we add `.. versionadded:: 3.16` to this member? The enclosing enum is documented as introduced in 3.13, so the new member's availability is otherwise unclear.

File: Python/compile.c [hunk 12]
```diff
@@ -907,17 +979,42 @@ compiler_mod(compiler *c, mod_ty mod)
...
+/* Inlined comprehensions are compiled in the enclosing unit. If a name is
+ * FREE in the comprehension, resolve it in enclosing tables until it is no
+ * longer FREE. Stop if the next table is a class: nested scopes (including
+ * inlined comprehensions) do not see class locals, so the name stays FREE.
+ * __class__ and friends are not allowed to be free through a class; treat
+ * those loads as implicit globals. */
+static int
+compiler_resolve_inlined_free(PySTEntryObject **ste, int scope, PyObject *name)
+{
+    while (scope == FREE && (*ste)->ste_type == InlinedComprehensionBlock) {
```
Comment (carljm) [inline Python/compile.c:991]: Could we account for the implicit receiver synthesized by `load_args_for_super()` here? Its scope is `0` in the comprehension's symbol table, so this `scope == FREE` condition skips the enclosing-scope lookup.

This now aborts compilation in a debug build:

```python
class A:
    def f(self):
        return 42

class B(A):
    def f(self):
        return [super().f() for _ in (0,)]

print(B().f())
```

Main returns `[42]`. The [implicit receiver load](https://github.com/python/cpython/blob/045091d3c60616a7124a2ba84660b01ab8e206d5/Python/codegen.c#L3951-L3957) passes `self` to `codegen_nameop()`, but `self` isn't explicitly referenced in the comprehension, so it has no symbol-table entry there. The unresolved scope reaches `codegen_nameop()`'s assertion. Renaming the receiver to `_self` gets past the assertion but raises `SystemError: no locals found` at runtime.

**Review by carljm — COMMENTED** [review]
This version looks like a major improvement! Codex did find a few issues.

**Comment by iritkatriel** [comment]
> Codex did find a few issues.

Impressive!

**Comment by carljm** [comment]
@iritkatriel I see all the inline comments are marked resolved, so I take it this is ready for another review?

**Comment by iritkatriel** [comment]
> @iritkatriel I see all the inline comments are marked resolved, so I take it this is ready for another review?

I think so. Thanks.

File: Python/symtable.c [hunk 13]
```diff
@@ -802,109 +827,113 @@ is_free_in_any_child(PySTEntryObject *entry, PyObject *key)
...
-            }
-            int ok = PyDict_SetItem(ste->ste_symbols, k, v_flags);
-            Py_DECREF(v_flags);
-            if (ok < 0) {
-                return 0;
-            }
-            SET_SCOPE(scopes, k, scope);
+        // Loads here are in the enclosing compilation unit.
+        if (scope == FREE && (comp_flags & USE) &&
+                !symtable_add_flag(ste->ste_symbols, k, USE)) {
```
Comment (carljm) [inline Python/symtable.c:908]: I think we should read the existing flags before this call, or hold an owned reference until after `PyLong_AsLong(existing)`. `existing` is borrowed from `ste->ste_symbols`, and `symtable_add_flag()` replaces that same dictionary value and decrefs the old integer. The subsequent `PyLong_AsLong(existing)` can therefore read an object whose lifetime has ended. Most flags fit in the small-integer cache, but type-parameter bindings can exceed it. It may be hard to reproduce an actual crash here, since the integer freelist could also mask the problem. But it still seems worth avoiding the theoretical use-after-free possibility here.

File: Objects/frameobject.c [hunk 14]
```diff
@@ -380,16 +396,28 @@ framelocalsproxy_keys(PyObject *self, PyObject *Py_UNUSED(ignored))
     if (names == NULL) {
         return NULL;
     }
+    // An inlined comprehension cell can share a name with a free var.
+    PyObject *seen = PySet_New(NULL);
```
Comment (carljm) [inline Objects/frameobject.c:400]: I wonder if we need to do any compilation performance testing of this PR on release builds? We are now doing this set allocation and per-name deduplication for every frame, including ordinary functions with no comprehensions, cells, or free variables. (Same for `values()`, `items()`, and `len()`.)

If this cost is totally invisible in real workloads, maybe this is fine and we just keep the code simpler. But it also seems quite feasible to limit the duplicate-tracking cost to code objects that could actually be affected by it.

File: InternalDocs/inlined_comprehensions.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+Since [PEP 709](https://peps.python.org/pep-0709/), list, set, and dict
+comprehensions are compiled into the enclosing compilation unit instead of
+creating a nested function and calling it. Generator expressions are not
+inlined; they still become their own code object.
+
+The resulting bytecode lives in the enclosing unit, but the comprehension
+still has its own locals: iteration variables must not leak into, or
+overwrite, names in the enclosing scope. The symbol table models that as a
+nested lexical scope; codegen then emits the comprehension in place.
```
Comment (carljm) [inline InternalDocs/inlined_comprehensions.md]: nit: is "in place" clear here? Would it be clearer to say "inlined into its containing compilation unit" or similar?

File: InternalDocs/inlined_comprehensions.md [hunk 16]
```diff
@@ -0,0 +1,131 @@
+Inlined comprehensions
```
Comment (carljm) [inline InternalDocs/inlined_comprehensions.md:1]: This is super helpful, thank you

**Review by carljm — APPROVED** [review]
This is excellent, thank you!!

File: Objects/frameobject.c (hunk 14 above)
Comment (iritkatriel) [reply Objects/frameobject.c:400]: Good point, though I think this code impacts introspection tools rather than the compiler. Let me see if it's easy to optimize.

Comment (iritkatriel) [reply Objects/frameobject.c:400]: Doable, but not trivial. I'd like to defer the optimisation to a separate PR.

**Comment by iritkatriel** [comment]
Many thanks to the people and bots. This was fun.
